frontend/src/LineGraph.res

```diff
@@ -153,49 +164,96 @@ let graphSx = [Sx.absolute, Sx.t.xl3, Sx.b.zero, Sx.l.zero, Sx.r.zero]
 let containerSx = [Sx.mt.xl2, Sx.relative, Sx.unsafe("width", "480px"), Sx.h.xl5]
 
 @react.component
-let make = (
+let make = React.memo((
   ~sx as uSx=[],
   ~title=?,
-  ~xTicks=?,
-  ~yLabel=?,
-  ~labels=?,
-  ~xLabelFormatter=?,
-  ~onRender: option<graph => unit>=?,
+  ~xTicks: option<Belt.Map.Int.t<string>>=?,
+  ~yLabel: option<string>=?,
+  ~labels: option<array<string>>=?,
   ~onXLabelClick=?,
-  ~annotations=[],
+  ~annotations: array<{
+    "clickHandler": (annotation, point, graph, event) => unit,
+    "height": int,
+    "icon": string,
+    "series": string,
+    "text": string,
+    "width": int,
+    "x": int,
+  }>=[],
   ~data,
 ) => {
-  let graphRef = React.useRef(Js.Nullable.null)
+  let graphDivRef = React.useRef(Js.Nullable.null)
+  let graphRef = React.useRef(None)
   React.useEffect1(() => {
     let options = defaultOptions(
       ~yLabel?,
       ~labels?,
       ~xTicks?,
-      ~xLabelFormatter?,
       ~legendFormatter=Legend.format(~xTicks?),
       (),
     )
-    let graph = init(graphRef.current, data, options)
-    switch onRender {
-    | Some(f) => f(graph)
+
+    switch Js.Nullable.toOption(graphDivRef.current) {
     | None => ()
+    | Some(ref) => {
+        let graph = init(ref, data, options)
+        graphRef.current = Some(graph)
+
+        if Array.length(annotations) > 0 {
+          graph->ready(() => {
+            graph->setAnnotations(annotations)
+          })
+        }
+
+        switch onXLabelClick {
+        | Some(handler) =>
+          getElementsByClassName("dygraph-axis-label-x")->Belt.Array.forEach(elem =>
+            getElementHTMLonClick(elem, handler)
+          )
+        | None => ()
+        }
+      }
     }
-    if Array.length(annotations) > 0 {
-      graph->ready(() => {
-        graph->setAnnotations(annotations)
-      })
-    }
 
-    switch onXLabelClick {
-    | Some(handler) =>
-      getElementsByClassName("dygraph-axis-label-x")->Belt.Array.forEach(elem =>
-        getElementHTMLonClick(elem, handler)
-      )
+    Some(
+      _ => {
+        switch graphRef.current {
+        | Some(graph) => {
+            graph->destroy
+            graphRef.current = None
+          }
+        | None => ()
+        }
+      },
+    )
+  }, [])
+
+  React.useLayoutEffect2(() => {
+    switch graphRef.current {
     | None => ()
-    }
+    | Some(graph) => {
+        let options = defaultOptions(
+          ~yLabel?,
+          ~labels?,
+          ~xTicks?,
+          ~data,
+          ~legendFormatter=Legend.format(~xTicks?),
+          (),
+        )
+        graph->updateOptions(options)
+        graph->setAnnotations(annotations)
 
+        switch onXLabelClick {
+        | Some(handler) =>
+          getElementsByClassName("dygraph-axis-label-x")->Belt.Array.forEach(elem =>
+            getElementHTMLonClick(elem, handler)
+          )
+        | None => ()
+        }
+      }
+    }
     None
-  }, [data])
+  }, (data, annotations))
```

Comment:
Ideally, the array should contain all the `defaultOptions` params so that the graph is updated if one of these options change. But in reality, I think these options only change when `data` change, so I guess that's fine as is.